pil-analyzer/tests/types.rs

```diff
@@ -561,6 +561,58 @@ fn empty_conditional() {
     type_check(input, &[]);
 }
 
+#[test]
+fn defined_trait() {
+    let input = "
+    trait Add<T> {
+        add: T, T -> T,
+    }
+    impl Add<fe> {
+        add: |a, b| a + b,
+    }
+    let r: fe = Add::add(3, 4);
+    ";
+    type_check(input, &[("r", "", "fe")]);
+}
+
+#[test]
+fn defined_trait_generic() {
+    let input = "
+    namespace std::convert(4);
+        let fe = || fe();
+    namespace F(4);
+        trait Add<T, Q> {
+            add: T, T -> Q,
+        }
+        impl<T> Add<T, fe> {
+            add: |a, b| std::convert::fe(a + b),
```

Comment:
This test is interesting because it should actually fail - but it is not related to trait unification / impl resolution:

You cannot call `std::convert::fe` on any type `T`. It has type constraints and these need to be matched to the ones in the `impl` here.